File: public/include/javascript/server/handler.js
goog.provide('server.handler');
goog.require('cljs.core');
server.handler.pull_out_signals = (function pull_out_signals(result_data,tag){
return cljs.core.map.call(null,(function (p1__101708_SHARP_){
return cljs.core.first.call(null,p1__101708_SHARP_);
}),cljs.core.remove.call(null,cljs.core.empty_QMARK_,cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.map.call(null,(function (inp){
return cljs.core.ObjMap.fromObject(["\uFDD0'x","\uFDD0'title","\uFDD0'text"],{"\uFDD0'x":window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),"\uFDD0'title":(new cljs.core.Keyword("\uFDD0'signal")).call(null,inp),"\uFDD0'text":[cljs.core.str("Why: "),cljs.core.str((new cljs.core.Keyword("\uFDD0'why")).call(null,inp))].join('')});
}),(new cljs.core.Keyword("\uFDD0'signals")).call(null,ech)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,tag.call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data)))))));
});
server.handler.pull_out_strategies = (function pull_out_strategies(result_data,tag){
var result_strategies = cljs.core.map.call(null,(function (p1__101709_SHARP_){
return cljs.core.first.call(null,p1__101709_SHARP_);
}),cljs.core.remove.call(null,cljs.core.empty_QMARK_,cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.map.call(null,(function (inp){
return cljs.core.ObjMap.fromObject(["\uFDD0'x","\uFDD0'title","\uFDD0'text"],{"\uFDD0'x":window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),"\uFDD0'title":(new cljs.core.Keyword("\uFDD0'signal")).call(null,inp),"\uFDD0'text":[cljs.core.str("Why: "),cljs.core.str((new cljs.core.Keyword("\uFDD0'why")).call(null,inp))].join('')});
}),(new cljs.core.Keyword("\uFDD0'strategies")).call(null,ech)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,tag.call(null,(new cljs.core.Keyword("\uFDD0'strategies")).call(null,result_data)))))));
return result_strategies;
});
server.handler.parse_result_data = (function parse_result_data(result_data){
return cljs.core.ObjMap.fromObject(["\uFDD0'local-list","\uFDD0'macd-price-list","\uFDD0'stock-name","\uFDD0'sma-list","\uFDD0'macd-histogram-list","\uFDD0'ema-list","\uFDD0'stochastic-d","\uFDD0'bollinger-band","\uFDD0'obv","\uFDD0'strategies","\uFDD0'stochastic-k","\uFDD0'macd-signal-list","\uFDD0'signals"],{"\uFDD0'local-list":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt(cljs.core.first.call(null,ech)),window.parseFloat(cljs.core.second.call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.into_array.call(null,(new cljs.core.Keyword("\uFDD0'stock-list")).call(null,result_data)))),"\uFDD0'macd-price-list":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'last-trade-macd")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'macd")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'stock-name":(new cljs.core.Keyword("\uFDD0'stock-name")).call(null,result_data),"\uFDD0'sma-list":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt(cljs.core.first.call(null,ech)),window.parseFloat(cljs.core.second.call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,(function (p1__101710_SHARP_){
return (cljs.core.first.call(null,p1__101710_SHARP_) == null);
}),cljs.core.into_array.call(null,(new cljs.core.Keyword("\uFDD0'sma-list")).call(null,result_data))))),"\uFDD0'macd-histogram-list":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'histogram")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'macd")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'ema-list":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt(cljs.core.first.call(null,ech)),window.parseFloat(cljs.core.second.call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,(function (p1__101711_SHARP_){
return (cljs.core.first.call(null,p1__101711_SHARP_) == null);
}),cljs.core.into_array.call(null,(new cljs.core.Keyword("\uFDD0'ema-list")).call(null,result_data))))),"\uFDD0'stochastic-d":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'D")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'stochastic-oscillator")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'bollinger-band":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'lower-band")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'upper-band")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'bollinger-band")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'obv":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseInt((new cljs.core.Keyword("\uFDD0'obv")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'obv")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'strategies":cljs.core.ObjMap.fromObject(["\uFDD0'strategy-A","\uFDD0'strategy-B"],{"\uFDD0'strategy-A":server.handler.pull_out_strategies.call(null,result_data,"\uFDD0'strategy-A"),"\uFDD0'strategy-B":server.handler.pull_out_strategies.call(null,result_data,"\uFDD0'strategy-B")}),"\uFDD0'stochastic-k":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'K")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'stochastic-oscillator")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'macd-signal-list":cljs.core.into_array.call(null,cljs.core.reduce.call(null,(function (rslt,ech){
return cljs.core.conj.call(null,rslt,cljs.core.into_array.call(null,cljs.core.PersistentVector.fromArray([window.parseInt((new cljs.core.Keyword("\uFDD0'last-trade-time")).call(null,ech)),window.parseFloat((new cljs.core.Keyword("\uFDD0'ema-signal")).call(null,ech))], true)));
}),cljs.core.PersistentVector.EMPTY,cljs.core.remove.call(null,cljs.core.nil_QMARK_,(new cljs.core.Keyword("\uFDD0'macd")).call(null,(new cljs.core.Keyword("\uFDD0'signals")).call(null,result_data))))),"\uFDD0'signals":cljs.core.ObjMap.fromObject(["\uFDD0'moving-average","\uFDD0'bollinger-band","\uFDD0'macd","\uFDD0'stochastic-oscillator","\uFDD0'obv"],{"\uFDD0'moving-average":server.handler.pull_out_signals.call(null,result_data,"\uFDD0'moving-average"),"\uFDD0'bollinger-band":server.handler.pull_out_signals.call(null,result_data,"\uFDD0'bollinger-band"),"\uFDD0'macd":server.handler.pull_out_signals.call(null,result_data,"\uFDD0'macd"),"\uFDD0'stochastic-oscillator":server.handler.pull_out_signals.call(null,result_data,"\uFDD0'stochastic-oscillator"),"\uFDD0'obv":server.handler.pull_out_signals.call(null,result_data,"\uFDD0'obv")})});
});

File: src/edgar/server/handler.clj
(ns edgar.server.handler)


(defn- pull-out-signals [result-data tag]

  (->> (reduce (fn [rslt ech]

                 (conj rslt (map (fn [inp]   ;; iterate over the :signals list, for each tick entry
                                   {:x (:last-trade-time ech)
                                    :title (:signal inp)
                                    :text (str "Why: " (:why inp))
                                    })
                                 (:signals ech))))
               []
               (remove nil? (-> result-data :signals tag)))
       (remove empty?)

       ;; for some strange reason, each list entry is in another list
       (map #(first %))))

(defn- pull-out-strategies [result-data tag]

  (let [result-strategies (->> (reduce (fn [rslt ech]

                                         #_(.log js/console (str "... pulling out strategies[" ech "]"))
                                         (conj rslt (map (fn [inp]
                                                           {:x (:last-trade-time ech)
                                                            :title (:signal inp)
                                                            :text (str "Why: " (:why inp))
                                                            })
                                                         (:strategies ech))))
                                       []
                                       (remove nil? (-> result-data :strategies tag)))
                               (remove empty?)

                               ;; for some strange reason, each list entry is in another list
                               (map #(first %)))]

    #_(.log js/console (str "... pulling out strategies > END[" result-strategies "]"))
    result-strategies))

(defn parse-result-data [result-data]

  {:local-list (reduce (fn [rslt ech]
                         (conj rslt [(first ech)
                                     (second ech)]))
                       []
                       (:stock-list result-data))


   ;; Basic Long and Short Moving Averages
   :sma-list (reduce (fn [rslt ech]
                       (conj rslt [(first ech)
                                   (second ech)]))
                     []
                     (remove #(nil? (first %))
                             (:sma-list result-data)))

   :ema-list (reduce (fn [rslt ech]
                       (conj rslt [(first ech)
                                   (second ech)]))
                     []
                     (remove #(nil? (first %))
                             (:ema-list result-data)))


   ;; Bollinger-Band Data
   :bollinger-band (reduce (fn [rslt ech]
                             (conj rslt [(:last-trade-time ech)
                                         (:lower-band ech)
                                         (:upper-band ech)]))
                           []
                           (remove nil? (-> result-data :signals :bollinger-band)))

   ;; MACD Data
   :macd-price-list (reduce (fn [rslt ech]
                              (conj rslt [(:last-trade-time ech)
                                          (:last-trade-macd ech)]))
                            []
                            (remove nil? (-> result-data :signals :macd)))

   :macd-signal-list (reduce (fn [rslt ech]
                               (conj rslt [(:last-trade-time ech)
                                           (:ema-signal ech)]))
                             []
                             (remove nil? (-> result-data :signals :macd)))

   :macd-histogram-list (reduce (fn [rslt ech]
                                  (conj rslt [(:last-trade-time ech)
                                              (:histogram ech)]))
                                []
                                (remove nil? (-> result-data :signals :macd)))

   ;; Stochastic Oscillator
   :stochastic-k (reduce (fn [rslt ech]
                           (conj rslt [(:last-trade-time ech)
                                       (:K ech)]))
                         []
                         (remove nil? (-> result-data :signals :stochastic-oscillator)))

   :stochastic-d (reduce (fn [rslt ech]
                           (conj rslt [(:last-trade-time ech)
                                       (:D ech)]))
                         []
                         (remove nil? (-> result-data :signals :stochastic-oscillator)))

   :obv (reduce (fn [rslt ech]
                  (conj rslt [(:last-trade-time ech)
                              (:obv ech)]))
                []
                (remove nil? (-> result-data :signals :obv)))

   :signals {:moving-average (pull-out-signals result-data :moving-average)
             :bollinger-band (pull-out-signals result-data :bollinger-band)
             :macd (pull-out-signals result-data :macd)
             :stochastic-oscillator (pull-out-signals result-data :stochastic-oscillator)
             :obv (pull-out-signals result-data :obv)}

   :strategies {:strategy-A (pull-out-strategies result-data :strategy-A)
                :strategy-B (pull-out-strategies result-data :strategy-B)}

   :stock-name (:stock-name result-data)})
